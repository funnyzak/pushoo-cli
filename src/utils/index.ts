import CmdExists from 'command-exists';
import fs, { lstatSync } from 'fs';
import YAML, { parse as yamlParse } from 'yaml';
import { resolve } from './promise';

export async function commandExists(commandString: string): Promise<boolean> {
  const [commandError] = await resolve(CmdExists(commandString));
  return commandError === undefined;
}

// write yaml file from json object
export async function writeYamlFile(
  filePath: string,
  data: any
): Promise<string> {
  const _yaml = YAML.stringify(data);
  await fs.writeFileSync(filePath, _yaml, 'utf8');
  return _yaml;
}

// read yaml file to json object
export async function readYamlFile<T>(
  filePath: string
): Promise<T> | undefined {
  if ((await fs.existsSync(filePath)) && (await lstatSync(filePath).isFile())) {
    return yamlParse(await fs.readFileSync(filePath, 'utf8')) as T;
  }
  return undefined;
}

// Check the object attributes required
export function checkRequiredAttributes(object: any, attributes: string[]) {
  let check_result = [];
  for (const attribute of attributes) {
    if (!object[attribute] || object[attribute] === null || object[attribute] === '') {
      check_result.push(attribute);
    }
  }
  if (check_result.length > 0) {
    throw new Error(`Missing required attributes: ${check_result.join(', ')}`);
  }
}

// Check whether the length of multiple array is the same
export function checkArrayLength(array: any[]) {
  let check_result = [];
  for (const item of array) {
    if (item.length !== array[0].length) {
      check_result.push(item);
    }
  }
  if (check_result.length > 0) {
    throw new Error(
      `The length of the array is not the same: ${check_result.join(', ')}`
    );
  }
}
