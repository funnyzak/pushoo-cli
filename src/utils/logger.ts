// A simple colorized console logger.

import chalk from 'chalk';

class LoggerManager {
  level = 'info';

  levels = ['trace', 'debug', 'info', 'warn', 'error'];

  constructor(level: string) {
    this.level = level;
  }

  setLevel(_level: string): void {
    this.level = _level;
  }

  shouldLog(log_level: string): boolean {
    return this.levels.indexOf(log_level) >= this.levels.indexOf(this.level);
  }

  trace(...message: string[]): void {
    if (this.shouldLog('trace')) {
      console.trace(chalk.bgGray.bold(' TRACE '), ...message);
    }
  }

  debug(...message: string[]): void {
    if (this.shouldLog('debug')) {
      console.debug(chalk.yellowBright.bold(' DEBUG '), ...message);
    }
  }

  info(...message: string[]): void {
    if (this.shouldLog('info')) {
      console.info(chalk.bgMagenta.bold(' INFO '), ...message);
    }
  }

  warn(...message: string[]): void {
    if (this.shouldLog('warn')) {
      console.error(chalk.bgYellow.bold(' WARN '), ...message);
    }
  }

  error(...message: string[]): void {
    if (this.shouldLog('error')) {
      console.error(chalk.bgRed.bold(' ERROR '), ...message);
    }
  }

  log(...message: string[]): void {
    console.log(...message);
  }
}

const logger = new LoggerManager('info');

export { logger };
