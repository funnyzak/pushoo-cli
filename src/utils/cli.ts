// CLI-related utility functions.

import parseArgv from 'arg';
import chalk from 'chalk';
import chalkTemplate from 'chalk-template';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { env } from 'node:process';
import prompt from 'prompt';
import checkForUpdate from 'update-check';
import { readYamlFile, writeYamlFile } from '.';
import type { Arguments } from '../types';
import { Prompts } from '../types';
import { logger } from './logger';
import { resolve } from './promise';

// The help text for the CLI.
const helpText = chalkTemplate`
{bold.cyan pushoo-cli} - a command line tool with Pushoo.js pushes multiple platform messages.

{bold Usage:}
  {cyan pushoo} [options]
  {cyan pushoo} --help
  {cyan pushoo} --version
  {cyan pushoo} "hello world"
  {cyan pushoo} -C "hello world"
  {cyan pushoo} -c ./pushoo.yml -C "hello world"
  {cyan pushoo} -P wecom -K wecom_token -C "hello world"

{bold Options:}
  {cyan -h, --help}                          Shows help.
  {cyan -d, --debug}                         Show debugging information.
  {cyan -v, --version}                       Print version of pushoo-cli.
  {cyan -c, --config}                        Optional, Specify the configuration file path.
  {cyan -C, --content}                       Required, The push content of the Markdown format.
  {cyan -P, --platforms}                     Optional, List of platform name( more: https://github.com/imaegoo/pushoo), such as: {underline wecom, dingtalk,feishu}.
 {cyan -K, --tokens}                        Optional, List of token( more: https://github.com/imaegoo/pushoo), such as: {underline wecom_token, dingtalk_token,feishu_token}.
  {cyan -T, --title}                         Optional, message title.
  {cyan -O, --options}                       Optional, For some additional configuration when pushing, Json string.

  By default, {cyan pushoo} If there is no corresponding parameter, read the configuration from the local configuration file to send.

{bold Commands:}
  {cyan config}                              Create configuration files to store the default configuration。
`;

/**
 * Returns the help text.
 *
 * @returns The help text shown when the `--help` option is used.
 */
export const getHelpText = (): string => helpText;

// The options the CLI accepts, and how to parse them.
const options = {
  '--help': Boolean,
  '--debug': Boolean,
  '--version': Boolean,
  '--config': String,
  '--content': String,
  '--platforms': String,
  '--tokens': String,
  '--title': String,
  '--options': String,

  // A list of aliases for the above options.
  '-h': '--help',
  '-d': '--debug',
  '-v': '--version',
  '-c': '--config',
  '-C': '--content',
  '-P': '--platforms',
  '-K': '--tokens',
  '-T': '--title',
  '-O': '--options'
};

/**
 * Parses the program's `process.argv` and returns the options and arguments.
 *
 * @returns The parsed options and arguments.
 */
export const parseArguments = (): Arguments => parseArgv(options);

/**
 * Checks for updates to this package. If an update is available, it brings it
 * to the user's notice by printing a message to the console.
 */
export const checkForUpdates = async (manifest: object): Promise<void> => {
  // Do not check for updates if the `NO_UPDATE_CHECK` variable is set.
  if (env.NO_UPDATE_CHECK) return;

  // Check for a newer version of the package.
  const [error, update] = await resolve(checkForUpdate(manifest));

  // If there is an error, throw it; and if there is no update, return.
  if (error) throw error;
  if (!update) return;

  // If a newer version is available, tell the user.
  logger.log(
    chalk.bgRed.white(' UPDATE '),
    `The latest version of \`pushoo-cli\` is ${update.latest}`
  );
};

// The default configuration file path.
export const defaultConfigurationFilePath = path.join(
  os.homedir(),
  '.pushoo.yml'
);


export const getConfiguration = (filepath: string) =>
  readYamlFile<Prompts.Configuration>(filepath);


// get default configuration json
export const getDefaultConfiguration = () =>
  getConfiguration(defaultConfigurationFilePath);

export const configurationFileSetting = async (): Promise<void> => {
  let _configuration = await readYamlFile<Prompts.Configuration>(
    defaultConfigurationFilePath
  );
  if (_configuration) {
    logger.log(
      chalkTemplate`The following is the current configuration: \n\n{cyan ${await fs.readFileSync(
        defaultConfigurationFilePath,
        'utf8'
      )}}`
    );

    const { confirm } = await prompt.get({
      properties: {
        confirm: {
          message: 'Reconfigure? (y/n)',
          required: false
        }
      }
    });

    if (confirm !== 'y') {
      return;
    }
  }

  _configuration = await prompt.get<Prompts.Configuration>({
    properties: {
      platforms: {
        message:
          'List of platform name( more: https://github.com/imaegoo/pushoo), such as: wecom, dingtalk, feishu',
        required: true,
        default: _configuration?.platforms
      },
      tokens: {
        message:
          'List of token( more: https://github.com/imaegoo/pushoo), such as: wecom_token, dingtalk_token, feishu_token',
        required: true,
        default: _configuration?.tokens
      },
      title: {
        message: 'Optional，Default message title',
        required: false,
        default: undefined
      },
      content: {
        message: 'optional, Default message content',
        required: false,
        default: undefined
      }
    }
  });

  if (
    _configuration.platforms.split(/[,，]/).length !==
    _configuration.tokens.split(/[,，]/).length
  ) {
    logger.error('The number of platforms and tokens must be the same');
    return;
  }

  const _yaml = await writeYamlFile(defaultConfigurationFilePath, _configuration);

  logger.log(
    chalkTemplate`The configuration content as follows：\n\n{cyan ${_yaml}}\nAnd configuration file has been stored：{cyan ${defaultConfigurationFilePath}}，You can modify the configuration through {cyan pushoo config}。`
  );
};
