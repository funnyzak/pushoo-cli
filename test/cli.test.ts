// Tests for the CLI part of the project.

import { tmpdir } from 'os';
import { env } from 'process';
import prompt from 'prompt';

import path from 'path';
import { afterEach, describe, expect, test, vi } from 'vitest';

import manifest from '../package.json';
import {
checkForUpdates, configurationFileSetting, defaultConfigurationFilePath, getHelpText
} from '../src/utils/cli';
import { logger } from '../src/utils/logger';

afterEach(() => {
  vi.restoreAllMocks();
});

describe('utils/cli', () => {
  // Make sure the help message remains the same. If we are changing the help
  // message, then make sure to run `vitest` with the `--update-snapshot` flag.
  test('render help text', () => expect(getHelpText()).toMatchSnapshot());

  test('default configuration path endwith yml', () =>
    expect(defaultConfigurationFilePath).toContain('pushoo.yml'));

  test.skip(
    'configuation file setting',
    async () => {
      const promptSpy = vi.spyOn(prompt, 'get');

      prompt.start();

      expect(promptSpy).not.toHaveBeenCalled();

      await configurationFileSetting(
        path.join(tmpdir(), `.pushoo_${new Date().getTime()}.yml`)
      );

      expect(promptSpy).toHaveBeenCalledOnce();
    },
    { timeout: 1000 }
  );
  // Make sure the update message is shown when the current version is not
  // the latest version.
  test('print update message when newer version exists', async () => {
    const consoleSpy = vi.spyOn(logger, 'log');

    await checkForUpdates({
      ...manifest,
      version: '0.0.0'
    });

    expect(consoleSpy).toHaveBeenCalledOnce();
    expect(consoleSpy).toHaveBeenLastCalledWith(
      expect.stringContaining('UPDATE'),
      expect.stringContaining('latest')
    );
  });

  // Make sure the update message is not shown when the latest version is
  // running.
  test('do not print update message when on latest version', async () => {
    const consoleSpy = vi.spyOn(logger, 'log');

    await checkForUpdates({
      ...manifest,
      version: '99.99.99'
    });

    expect(consoleSpy).not.toHaveBeenCalled();
  });

  // Make sure an update check does not occur when the NO_UPDATE_CHECK env var
  // is set.
  test('do not check for updates when NO_UPDATE_CHECK is set', async () => {
    const consoleSpy = vi.spyOn(logger, 'log');

    env.NO_UPDATE_CHECK = 'true';
    await checkForUpdates({
      ...manifest,
      version: '0.0.0'
    });
    env.NO_UPDATE_CHECK = undefined;

    expect(consoleSpy).not.toHaveBeenCalled();
  });
});
