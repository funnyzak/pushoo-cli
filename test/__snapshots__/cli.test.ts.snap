// Vitest Snapshot v1

exports[`utils/cli > render help text 1`] = `
"
[1m[36mpushoo-cli[39m[22m - a command line tool with Pushoo.js pushes multiple platform messages.

[1mUsage:[22m
  [36mpushoo[39m [options]
  [36mpushoo[39m --help
  [36mpushoo[39m --version
  [36mpushoo[39m \\"hello world\\"
  [36mpushoo[39m -C \\"hello world\\"
  [36mpushoo[39m -c ./pushoo.yml -C \\"hello world\\"
  [36mpushoo[39m -P wecom -K wecom_token -C \\"hello world\\"

[1mCommands:[22m
  [36mconfig[39m                              Create default configuration file.

[1mOptions:[22m
  [36m-h, --help[39m                          Shows help.
  [36m-d, --debug[39m                         Show debugging information.
  [36m-v, --version[39m                       Print version of pushoo-cli.
  [36m-c, --config[39m                        Optional, Specify the configuration file path.
  [36m-C, --content[39m                       Required, The push content of the Markdown format.
  [36m-P, --platforms[39m                     Optional, List of platform name(more: https://github.com/imaegoo/pushoo), such as: [4mwecom, dingtalk,feishu[24m.
  [36m-K, --tokens[39m                        Optional, List of token(more: https://github.com/imaegoo/pushoo), such as: [4mwecom_token, dingtalk_token,feishu_token[24m.
  [36m-T, --title[39m                         Optional, message title.
  [36m-O, --options[39m                       Optional, For some additional configuration when pushing, Json string.

  By default, [36mpushoo[39m If there is no corresponding parameter, read the configuration from the local configuration file to send.

  More information about the pushoo-cli can be found at: [36m[4mhttps://github.com/funnyzak/pushoo-cli[24m[39m, and the pushoo can be found at: [36m[4mhttps://github.com/imaegoo/pushoo[24m[39m.
"
`;
